merkle/map_verifier.go

```diff
@@ -30,32 +29,26 @@ import (
 //
 // Returns nil on a successful verification, and an error otherwise.
 func VerifyMapInclusionProof(treeID int64, index, leafHash, expectedRoot []byte, proof [][]byte, h hashers.MapHasher) error {
-	hBits := h.Size() * 8
-
-	if got, want := len(proof), hBits; got != want {
-		return fmt.Errorf("proof len: %d, want %d", got, want)
-	}
-	if got, want := len(index)*8, hBits; got != want {
+	if got, want := len(index)*8, h.BitLen(); got != want {
 		return fmt.Errorf("index len: %d, want %d", got, want)
 	}
-	if got, want := len(leafHash)*8, hBits; got != want {
-		return fmt.Errorf("leafHash len: %d, want %d", got, want)
+	if got, want := len(proof), h.BitLen(); got != want {
+		return fmt.Errorf("proof len: %d, want %d", got, want)
+	}
+	for i, element := range proof {
+		if got, wanta, wantb := len(element), 0, h.Size(); got != wanta && got != wantb {
+			return fmt.Errorf("proof[%d] len: %d, want %d or %d", i, got, wanta, wantb)
+		}
 	}
 
-	// TODO(al): Remove this dep on storage, since clients will want to use this code.
-	nID := storage.NewNodeIDFromHash(index)
-
 	runningHash := make([]byte, len(leafHash))
 	copy(runningHash, leafHash)
 
-	for bit := 0; bit < hBits; bit++ {
-		proofIsRightHandElement := nID.Bit(bit) == 0
-		pElement := proof[bit]
+	for height := 0; height < h.BitLen(); height++ {
```

Comment:
There are still some cases where we've used 'depth' for this elsewhere. I can't see anywhere else that uses 'height' for coordinates. We've tried to switch to 'level' e.g. in the path code because this is defined in our tree coordinate space where 0 is leaf level and it increases upwards.